entity-framework/core/testing/choosing-a-testing-strategy.md

```diff
@@ -0,0 +1,112 @@
+
+Since only the `DbSet` itself is stubbed and the query is evaluated in-memory, this approach ends up being very similar to using the EF Core InMemory provider: both techniques execute query operators in .NET over an in-memory collection. As a result, this technique suffers from the same drawbacks as well: queries will behave differently (e.g. around case sensitivity) or will simply fail (e.g. because of provider-specific methods), raw SQL won't work and transactions will be ignored at best. As a result, this technique should generally be avoided for testing any query code.
+
+### Repository pattern
+
+The approaches above attempted to either swap EF Core's production database provider with a fake testing provider, or to stub out `DbSet` itself with an in-memory collection. These techniques are similar in that they evaluate the program's LINQ queries - either in SQLite or in memory - and this is ultimately the source of the difficulties outlined above: a query designed to run against a specific production database cannot run elsewhere without issues.
+
+For proper, reliable unit testing, consider introducing a [repository layer](https://martinfowler.com/eaaCatalog/repository.html) which mediates between your application code and EF Core. The production implementation of the repository contains the actual LINQ queries and executes them via EF Core. In testing, the repository abstraction is directly stubbed or mocked without needing any actual LINQ queries, effectively removing EF Core from your testing stack altogether and allowing tests to focus on application code alone.
+
+The following diagram compares the database fake pattern (SQLite/InMemory) with the repository pattern:
+
+![Comparison of fake provider with repository pattern](_static/fake-provider-and-repository-pattern.png)
+
+Since LINQ queries are no longer part of testing, you can directly provide query results to your application. Put another way, the previous approaches roughly allow stubbing out *query inputs* (e.g. replacing SQL Server tables with an in-memory one), but then still execute the actual query operators in-memory; the repository pattern, in contrast, allows you to stub out *query outputs* directly, allowing for far more powerful and focused unit testing.
+
+One possible disadvantage of the repository pattern is that it requires you to architect your application accordingly: application code can no longer execute LINQ queries directly, but must instead call into a repository method, which executes the query (and is stubbed in testing). This can make for a somewhat heavier application architecture, but can also have other advantages: all data access code is concentrated in one place rather than being spread around the application, and if your application needs to support more than one database, the repository abstraction can be very helpful for tweaking queries across providers.
```

Comment:
I think it's important to call out even more explicitly the disadvantages of a repository pattern. Maybe something like:

> LINQ is a very powerful feature of .NET. It allows applications to compose complex queries that are translated and executed against the database without the need to shape the data layer to explicitly support these queries. In addition to being powerful, this also provides a great deal of flexibility for the evolution of applications over time.

> Using a repository pattern to facilitate unit testing as described here necessarily prevents applications from using LINQ in this way, since if the repository itself exposes LINQ, then the testing problem has not been solved. This means that there is a significant cost involved in implementing and maintaining a repository. This cost should not be discounted when making a choice on how to test an application, especially given that integration tests for the queries generated by the repository will still be needed anyway.

> It's worth noting that repositories do have advantages outside of just testing. They ensure all data access code is concentrated in one place rather than being spread around the application, and if your application needs to support more than one database, then the repository abstraction can be very helpful for tweaking queries across providers.

Reply:
Also worth mentioning that leaking `IQueryable` while sometimes frowned upon, allows querying to become a minor cross-cutting concern.  To me, this is actually quite powerful, especially given that `IQueryable` is part of the language itself, and so leaking it doesn't create any unwanted dependencies.

( https://chillicream.com/blog/2020/03/18/entity-framework )

Reply:
@atrauzzi the moment you leak IQueryable, you lose the ability to stub/mock your repository layer - which at least in this context is the main reason for having a repository in the first place... I get what you're saying: if you have a complex query used in multiple places, and you want to compose further operators over it in those places, then yeah, you can wrap it in a method returning IQueryable for further composition. But if you want to preserve the ability to stub out your repository, each use of that query itself must be implemented in a method which *doesn't* return IQueryable, so that it can be stubbed/mocked.

Bottom line, I think this would be more confusing than helpful in this specific conversation, which is about testing/stubbing/mocking (GraphQL is a very specific case which again is somewhat orthogonal to testing).

Reply:
@ajcvickers I get what you're saying: it's indeed important to stress that testable repositories can't return IQueryable. But  I wouldn't say a repository pattern prevents composing complex queries with different result shapes, or any other composable operators (I think that's what you're saying). The basic limitation is that whatever API your product code uses must itself not return IQueryable; but a repository can have a private Queryable-returning method representing a complex query, and multiple public methods which compose over that (projections, filters, whatever) to return IEnumerable.

You could definitely argue that this is more laborious, as each data shape requires another IEnumerable-returning repository method, but that's just the regular price of the pattern (and different from saying that a repository prevents anything, no?)

I've integrated parts of your suggested text above, and hopefully made it clearer that the repository pattern has a significant implementation/maintenance cost - let me know what you think, am open for further changes.

Reply:
@atrauzzi I'm in favor of using IQueryable throughout application code. Hence I'm not in favor of the repository pattern. Something that looks like a repository but exposes IQueryable...is DbContext/DbSet. :-)

@roji

> But I wouldn't say a repository pattern prevents composing complex queries with different result shapes, or any other composable operators

I'm explicitly not saying this. I'm saying you have to write and maintain code in the repository to do this. There are good ways to do this. However, they are not IQueryable, and are (necessarily and by-design) more restrictive than IQueryable. This is what you lose/gain from having the repository; not the ability to have parameterized/extendable queries.